File: lib/src/widget.dart
// Copyright (c) 2016, Herman Bergwerf. All rights reserved.
// Use of this source code is governed by a MIT-style license
// that can be found in the LICENSE file.

part of fltk;

/// Widget callback function
typedef void Callback(Widget target, dynamic userData);

/// [Widget.onCallback] data
class WidgetCallbackData {
  final Widget target;
  final userData;
  WidgetCallbackData(this.target, this.userData);
}

/// [Widget.onResize] data
class WidgetResizeData {
  final int x, y, w, h;
  WidgetResizeData(this.x, this.y, this.w, this.h);
}

/// Fl_Widget
class Widget extends NativeFieldWrapperClass2 {
  /// User data
  dynamic userData;

  /// Callback function (classic style)
  Callback callback;

  /// Callback stream (Dart style)
  Stream<WidgetCallbackData> onCallback;

  /// Callback stream controller
  final _onCallbackController =
      new StreamController<WidgetCallbackData>.broadcast(sync: useSyncStreams);

  /// Resize stream
  Stream<WidgetResizeData> onResize;

  /// Resize stream controller
  ///
  /// Note that this is a synchronous stream because to prevent resize and draw
  /// events from mixing up.
  final _onResizeController =
      new StreamController<WidgetResizeData>.broadcast(sync: true);

  /// Create Fl_Widget (uses a wrapper class under the hood).
  Widget(int x, int y, int w, int h, [String l = '']) {
    _createWidget(x, y, w, h, l);
    _setupStreams();
  }

  Widget.empty() {
    _setupStreams();
  }

  /// Setup streams (call on create).
  void _setupStreams() {
    onCallback = _onCallbackController.stream;
    onResize = _onResizeController.stream;
  }

  /// Handle event (called by [doHandle]).
  bool handle(Event event) => false;

  /// Handle event code (called by the extension using `Dart_Invoke`).
  int doHandle(int event) => handle(Event.values[event]) ? 1 : 0;

  /// Draw widget (called by the extension using `Dart_Invoke`)
  void draw() {}

  /// Run callbacks (called by the extension using `Dart_Invoke`)
  void doCallback() {
    _onCallbackController.add(new WidgetCallbackData(this, userData));
    if (callback != null) {
      callback(this, userData);
    }
  }

  /// The widget is resized (called by the extension using `Dart_Invoke`).
  /// If you override this method you should call `super.resize` to keep the
  /// [onResize] stream working.
  void resize(int x, int y, int w, int h) {
    _onResizeController.add(new WidgetResizeData(x, y, w, h));
  }

  /// Native constructor
  void _createWidget(int x, int y, int w, int h, String l)
      native 'fldart::Widget::constructor_Widget';

  /// Get widget x position.
  int x() native 'fldart::Widget::int_x';

  /// Get widget y position.
  int y() native 'fldart::Widget::int_y';

  /// Get widget width.
  int w() native 'fldart::Widget::int_w';

  /// Get widget height.
  int h() native 'fldart::Widget::int_h';

  /// Activate the widget.
  void activate() native 'fldart::Widget::void_activate';

  /// Deactivate the widget.
  void deactivate() native 'fldart::Widget::void_deactivate';

  /// Get label.
  String get label native 'fldart::Widget::String_label';

  /// Set label.
  set label(String text) native 'fldart::Widget::void_label';

  /// Get label font.
  int get labelfont native 'fldart::Widget::Fl_Font_labelfont';

  /// Set label font.
  set labelfont(int font) native 'fldart::Widget::void_labelfont';

  /// Get label size.
  int get labelsize native 'fldart::Widget::Fl_Fontsize_labelsize';

  /// Set label size.
  set labelsize(int pixels) native 'fldart::Widget::void_labelsize';

  int _labeltypeGet() native 'fldart::Widget::Fl_Labeltype_labeltype';

  void _labeltypeSet(int type) native 'fldart::Widget::void_labeltype';

  /// Get label type.
  Labeltype get labeltype => Labeltype.values[_labeltypeGet()];

  /// Set label type.
  set labeltype(Labeltype type) => _labeltypeSet(type.index);

  /// Get label color.
  int get labelcolor native 'fldart::Widget::Fl_Color_labelcolor';

  /// Set label color
  set labelcolor(int color) native 'fldart::Widget::void_labelcolor';

  /// Get background color.
  int get color native 'fldart::Widget::Fl_Color_color';

  /// Set background color.
  set color(int color) native 'fldart::Widget::void_color';

  /// Get callback flags.
  int get when native 'fldart::Widget::int_when';

  /// Set callback flags.
  set when(int mode) native 'fldart::Widget::void_when';

  void _boxSet(int type) native 'fldart::Widget::void_box';

  /// Set widget box style.
  set box(Boxtype type) => _boxSet(type.index);

  /// Show widget.
  void show() native 'fldart::Widget::void_show';

  /// Hide widget.
  void hide() native 'fldart::Widget::void_hide';

  /// Redraw widget.
  void redraw() native 'fldart::Widget::void_redraw';

  void _image(int width, int height, int depth, Uint8List data)
      native 'fldart::Widget::void_image';

  /// Change widget background image.
  set image(Image image) => _image(image.width, image.height, 4,
      new Uint8List.fromList(image.data.buffer.asUint8List()));
}

File: ext/src/widgets/Widget.cpp
// Copyright (c) 2016, Herman Bergwerf. All rights reserved.
// Use of this source code is governed by a MIT-style license
// that can be found in the LICENSE file.

#include "Widget.hpp"

namespace fldart {
FunctionMapping Widget::methods[] = {
  {"fldart::Widget::x", Widget::x},
  {"fldart::Widget::y", Widget::y},
  {"fldart::Widget::w", Widget::w},
  {"fldart::Widget::h", Widget::h},
  {"fldart::Widget::show", Widget::show},
  {"fldart::Widget::box", Widget::box},
  {"fldart::Widget::label", Widget::label},
  {"fldart::Widget::labelfont", Widget::labelfont},
  {"fldart::Widget::labelsize", Widget::labelsize},
  {"fldart::Widget::labeltype", Widget::labeltype},
  {NULL, NULL}
};

void Widget::x(Dart_NativeArguments arguments) {
  int64_t ptr;
  Fl_Widget* _ref;
  Dart_EnterScope();
  HandleError(Dart_IntegerToInt64(HandleError(Dart_GetNativeArgument(arguments, 0)), &ptr));
  _ref = (Fl_Widget*)ptr;
  int64_t _tmp = _ref -> x();
  Dart_Handle _ret = Dart_NewInteger(_tmp);
  Dart_SetReturnValue(arguments, _ret);
  Dart_ExitScope();
}

void Widget::y(Dart_NativeArguments arguments) {
  int64_t ptr;
  Fl_Widget* _ref;
  Dart_EnterScope();
  HandleError(Dart_IntegerToInt64(HandleError(Dart_GetNativeArgument(arguments, 0)), &ptr));
  _ref = (Fl_Widget*)ptr;
  int64_t _tmp = _ref -> y();
  Dart_Handle _ret = Dart_NewInteger(_tmp);
  Dart_SetReturnValue(arguments, _ret);
  Dart_ExitScope();
}

void Widget::w(Dart_NativeArguments arguments) {
  int64_t ptr;
  Fl_Widget* _ref;
  Dart_EnterScope();
  HandleError(Dart_IntegerToInt64(HandleError(Dart_GetNativeArgument(arguments, 0)), &ptr));
  _ref = (Fl_Widget*)ptr;
  int64_t _tmp = _ref -> w();
  Dart_Handle _ret = Dart_NewInteger(_tmp);
  Dart_SetReturnValue(arguments, _ret);
  Dart_ExitScope();
}

void Widget::h(Dart_NativeArguments arguments) {
  int64_t ptr;
  Fl_Widget* _ref;
  Dart_EnterScope();
  HandleError(Dart_IntegerToInt64(HandleError(Dart_GetNativeArgument(arguments, 0)), &ptr));
  _ref = (Fl_Widget*)ptr;
  int64_t _tmp = _ref -> h();
  Dart_Handle _ret = Dart_NewInteger(_tmp);
  Dart_SetReturnValue(arguments, _ret);
  Dart_ExitScope();
}

void Widget::show(Dart_NativeArguments arguments) {
  int64_t ptr;
  Fl_Widget* _ref;
  Dart_EnterScope();
  HandleError(Dart_IntegerToInt64(HandleError(Dart_GetNativeArgument(arguments, 0)), &ptr));
  _ref = (Fl_Widget*)ptr;
  _ref -> show();
  Dart_Handle _ret = Dart_Null();
  Dart_SetReturnValue(arguments, _ret);
  Dart_ExitScope();
}

void Widget::box(Dart_NativeArguments arguments) {
  int64_t ptr;
  Fl_Widget* _ref;
  int64_t type;
  Dart_EnterScope();
  HandleError(Dart_IntegerToInt64(HandleError(Dart_GetNativeArgument(arguments, 0)), &ptr));
  _ref = (Fl_Widget*)ptr;
  HandleError(Dart_IntegerToInt64(HandleError(Dart_GetNativeArgument(arguments, 1)), &type));
  _ref -> box(static_cast<Fl_Boxtype>(type));
  Dart_Handle _ret = Dart_Null();
  Dart_SetReturnValue(arguments, _ret);
  Dart_ExitScope();
}

void Widget::label(Dart_NativeArguments arguments) {
  int64_t ptr;
  Fl_Widget* _ref;
  const char* text;
  Dart_EnterScope();
  HandleError(Dart_IntegerToInt64(HandleError(Dart_GetNativeArgument(arguments, 0)), &ptr));
  _ref = (Fl_Widget*)ptr;
  HandleError(Dart_StringToCString(HandleError(Dart_GetNativeArgument(arguments, 1)), &text));
  _ref -> label(newstr(text));
  Dart_Handle _ret = Dart_Null();
  Dart_SetReturnValue(arguments, _ret);
  Dart_ExitScope();
}

void Widget::labelfont(Dart_NativeArguments arguments) {
  int64_t ptr;
  Fl_Widget* _ref;
  int64_t f;
  Dart_EnterScope();
  HandleError(Dart_IntegerToInt64(HandleError(Dart_GetNativeArgument(arguments, 0)), &ptr));
  _ref = (Fl_Widget*)ptr;
  HandleError(Dart_IntegerToInt64(HandleError(Dart_GetNativeArgument(arguments, 1)), &f));
  _ref -> labelfont(f);
  Dart_Handle _ret = Dart_Null();
  Dart_SetReturnValue(arguments, _ret);
  Dart_ExitScope();
}

void Widget::labelsize(Dart_NativeArguments arguments) {
  int64_t ptr;
  Fl_Widget* _ref;
  int64_t pix;
  Dart_EnterScope();
  HandleError(Dart_IntegerToInt64(HandleError(Dart_GetNativeArgument(arguments, 0)), &ptr));
  _ref = (Fl_Widget*)ptr;
  HandleError(Dart_IntegerToInt64(HandleError(Dart_GetNativeArgument(arguments, 1)), &pix));
  _ref -> labelsize(pix);
  Dart_Handle _ret = Dart_Null();
  Dart_SetReturnValue(arguments, _ret);
  Dart_ExitScope();
}

void Widget::labeltype(Dart_NativeArguments arguments) {
  int64_t ptr;
  Fl_Widget* _ref;
  int64_t type;
  Dart_EnterScope();
  HandleError(Dart_IntegerToInt64(HandleError(Dart_GetNativeArgument(arguments, 0)), &ptr));
  _ref = (Fl_Widget*)ptr;
  HandleError(Dart_IntegerToInt64(HandleError(Dart_GetNativeArgument(arguments, 1)), &type));
  _ref -> labeltype(static_cast<Fl_Labeltype>(type));
  Dart_Handle _ret = Dart_Null();
  Dart_SetReturnValue(arguments, _ret);
  Dart_ExitScope();
}
}
